concurrency-primer.tex

```diff
@@ -452,52 +473,78 @@ \subsection{Compare and swap}
         return true;
+    } else {
+        memcpy(expected, obj, sizeof(*object));
+        return false;
     }
-    expected = *this;
-    return false;
 }
-\end{cppcode}
+\end{ccode}
 
 \begin{samepage}
 \noindent The \cpp|_strong| suffix may leave you wondering if there is a corresponding ``weak'' \textsc{CAS}.
 Indeed, there is. However, we will delve into that topic later in \secref{spurious-llsc-failures}.
 \end{samepage}
 
-Let's say we have some long-running task that we might want to cancel.
-We'll give it three states: \textit{idle}, \textit{running},
-and \textit{cancelled}, and write a loop that exits when it is cancelled.
-\begin{cppcode}
-enum class TaskState : int8_t {
-    Idle, Running, Cancelled
-};
-
-std::atomic<TaskState> ts;
-
-void taskLoop()
-{
-    ts = TaskState::Running;
-    while (ts == TaskState::Running) {
-      // Do good work.
-    }
-}
-\end{cppcode}
-If we want to cancel the task if it is running, but do nothing if it is idle,
-we could \textsc{CAS}:
-\begin{cppcode}
-bool cancel()
-{
-    auto expected = TaskState::Running;
-    return ts.compare_exchange_strong(expected, TaskState::Cancelled);
+Once threads are created in the thread pool as workers, they will continuously search for jobs to do.
+Jobs are taken from the tail of job queue.
+To claim a job without it being taken by another worker halfway through, we need to atomically change the pointer to the last job. Otherwise the last job is under races.
+The while loop in function \monobox{worker},
+\begin{ccode}
+while (!atomic_compare_exchange_weak(&thrd_pool->head->prev, &job,
+                                       job->prev)) {
 }
-\end{cppcode}
+\end{ccode}
+, keeps trying to claim the job atomically until success.
+
+Built-in post increment and decrement operators and compound assignment on atomic objects, such as \monobox{++} and \monobox{+=}, are read-modify-write atomic operations with total sequentially consistent ordering as well. 
+They behave equivalently to a \cc|do while| loop. See \clang{}11 standard 6.5.2.4 and 6.5.16.2 for more details.
+
+What if claiming a job, which updates \cc|thrd_pool->head->prev|, is not done atomically?
+Let's remove the atomicity of claiming a job and use thread sanitizer to detect races.
+Thread sanitizer is one of the sanitizers provided by \introduce{gcc} and \introduce{clang} to detect data races. 
+Data races are undefined behavior in \clang{}11 and \cplusplus{}11. 
+Thread sanitizer inserts runtime code into the target program to track memory accesses. 
+When races occur during execution, warning messages are printed.
+To enable this, add \monobox{-fsanitize=thread -g} to compiler flags.
+The following diff patch removes the atomicity of claiming a job and uses pthread instead of \clang{}11 thread, because thread sanitizer currently hasn't support \clang{}11 thread yet. 
+Save diff as \monobox{racer.diff} and patch the example code by \monobox{\$ patch rmw\_example.c race.diff}.
```

Comment:
Don't do that. This is a book, and we should emphasize the readability.

Reply:
Got it. Would it be a better way to use conditional macros in the example code? like this:
```c
typedef struct tpool {
    atomic_flag initialezed;
    int size;
    atomic_int state;
#ifdef PTHREAD
    pthread_t *pool;
#else
    thrd_t *pool;
#endif
    idle_job_t *head;
} tpool_t;
```
then use a short a description guiding the reader to edit 3 lines of code to create race condition.

Reply:
Let's stick to C11 Threads rather than POSIX Thread.

Reply:
Resolved in dcb8f9a.